File: server/app/services/scryfall.py
import httpx


class ScryfallService:
    def __init__(self) -> None:
        self.client = httpx.AsyncClient(
            headers={
                "Accept": "application/json",
                "User-Agent": "mtg-app/voice-card-flash",
            },
            timeout=10.0,
        )

    async def close(self) -> None:
        await self.client.aclose()

    @staticmethod
    def _get_image(card: dict) -> str | None:
        uris = card.get("image_uris") or {}
        if uris.get("normal"):
            return uris["normal"]
        if uris.get("large"):
            return uris["large"]
        faces = card.get("card_faces") or []
        if faces:
            face_uris = faces[0].get("image_uris") or {}
            if face_uris.get("normal"):
                return face_uris["normal"]
            if face_uris.get("large"):
                return face_uris["large"]
        return None

    async def lookup(self, query: str) -> dict:
        """
        Returns card dict on success.
        Raises httpx.HTTPStatusError on non-2xx from Scryfall.
        Raises ValueError if card has no image.
        """
        response = await self.client.get(
            "https://api.scryfall.com/cards/named",
            params={"fuzzy": query},
        )
        response.raise_for_status()
        card = response.json()
        image_url = self._get_image(card)
        if not image_url:
            raise ValueError("Card image unavailable")
        return {
            "id": card["id"],
            "name": card["name"],
            "typeLine": card["type_line"],
            "imageUrl": image_url,
        }

File: src/lib/scryfall.ts
const SCRYFALL_BASE = 'https://api.scryfall.com'

async function fetchWithTimeout(url: string, timeoutMs = 10000): Promise<Response> {
  const controller = new AbortController()
  const timer = setTimeout(() => controller.abort(), timeoutMs)
  try {
    const res = await fetch(url, { signal: controller.signal })
    return res
  } finally {
    clearTimeout(timer)
  }
}

export async function autocomplete(query: string): Promise<string[]> {
  if (!query || query.trim().length < 2) return []
  const url = `${SCRYFALL_BASE}/cards/autocomplete?q=${encodeURIComponent(query)}`
  const res = await fetchWithTimeout(url)
  if (!res.ok) return []
  const data = await res.json()
  return Array.isArray(data.data) ? data.data : []
}

export interface ScryfallCard {
  id: string
  name: string
  typeLine: string
  imageUrl: string
  colorIdentity: string[]
}

export async function lookup(query: string): Promise<ScryfallCard> {
  const url = `${SCRYFALL_BASE}/cards/named?fuzzy=${encodeURIComponent(query)}`
  const res = await fetchWithTimeout(url)
  if (!res.ok) {
    throw new Error(`Scryfall lookup failed: ${res.status}`)
  }
  const card = await res.json()

  let imageUrl = ''
  if (card.image_uris?.normal) {
    imageUrl = card.image_uris.normal
  } else if (card.image_uris?.large) {
    imageUrl = card.image_uris.large
  } else if (card.card_faces?.[0]?.image_uris?.normal) {
    imageUrl = card.card_faces[0].image_uris.normal
  }

  return {
    id: card.id,
    name: card.name,
    typeLine: card.type_line || '',
    imageUrl,
    colorIdentity: Array.isArray(card.color_identity) ? card.color_identity : [],
  }
}
